### PR TITLE
feat(compiler): Structural typing for data and abilities [fixes LNG-215]

### PR DESCRIPTION
**Pull request by DieMyst**

### CODE REVIEW

File: model/inline/src/main/scala/aqua/model/inline/ArrowInliner.scala [hunk 1]
```diff
@@ -313,22 +313,28 @@ object ArrowInliner extends Logging {
...
+                arrows
+                  .get(name)
+                  .map(fa =>
+                    (
+                      Map(currentNewName.getOrElse(currentOldName) -> vm),
+                      Map(name -> fa)
+                    )
                   )
-                )
+              case lm @ LiteralModel(_, _) =>
```
Comment (InversionSpaces) [inline model/inline/src/main/scala/aqua/model/inline/ArrowInliner.scala:326]: Why not something like
```
case model =>
  logger.error(s"Unexpected, $model can not be an arrow")
  None
```
?

File: model/inline/src/main/scala/aqua/model/inline/state/Exports.scala [hunk 2]
```diff
@@ -44,9 +44,9 @@ trait Exports[S] extends Scoped[S] {
   def copyWithAbilityPrefix(prefix: String, newPrefix: String): State[S, Unit]
 
   /**
-   * Get name of last linked VarModel
+   * Get name of last linked VarModel. If the last element will be LiteralModel, return None
```
Comment (InversionSpaces) [inline model/inline/src/main/scala/aqua/model/inline/state/Exports.scala:47]:
```suggestion
   * Get name of last linked VarModel. If the last element is not VarModel, return None
```

File: types/src/main/scala/aqua/types/CompareTypes.scala [hunk 3]
```diff
@@ -63,19 +63,21 @@ object CompareTypes {
...
-      ) == 1.0
-    ) 1.0
-    else if (
-      rf.keys.forall(lf.contains) && compareTypesList(
-        lfView.filterKeys(rfNEM.keys.contains).values.toList,
-        rfView.values.toList
-      ) == -1.0
-    ) -1.0
-    else NaN
+    else if (lf.keys.forall(rf.contains)) {
```
Comment (InversionSpaces) [inline types/src/main/scala/aqua/types/CompareTypes.scala:66]: What about the opposite case? 
When `rf.keys.forall(lf.contains)`

File: types/src/main/scala/aqua/types/CompareTypes.scala [hunk 4]
```diff
@@ -63,19 +63,21 @@ object CompareTypes {
...
+      if (
+        compareTypesList(
+          lfView.values.toList,
+          rfView.filterKeys(lfNEM.keys.contains).values.toList
+        ) >= 0.0
+      ) 1.0
+      else if (
+        compareTypesList(
```
Comment (InversionSpaces) [inline types/src/main/scala/aqua/types/CompareTypes.scala:74]: Is this call correct?
We know that `rf` contains all `lf` fields, but not the opposite

File: model/inline/src/main/scala/aqua/model/inline/ArrowInliner.scala (hunk 1 above)
Comment (DieMyst) [reply model/inline/src/main/scala/aqua/model/inline/ArrowInliner.scala:326]: because if there will be another model, we will handle it separately

**Review by InversionSpaces — APPROVED** [review]

**Pull request merged**